tree/tree/src/TTree.cxx

```diff
@@ -6580,7 +6580,12 @@ void TTree::OptimizeBaskets(ULong64_t maxMemory, Float_t minComp, Option_t *opti
          if (bsize < 0) bsize = bmax;
          if (bsize > bmax) bsize = bmax;
          UInt_t newBsize = UInt_t(bsize);
-         newBsize = newBsize - newBsize%512;
+         if (pass) { // only on the second pass so that it doesn't interfere with scaling
+            if (branch->GetBasket(0) != 0) {
+               newBsize = newBsize + (branch->GetBasket(0)->GetNevBuf() * sizeof(Int_t) * 2); // make room for meta dat
+            }
+            newBsize = newBsize - newBsize%512 + 512; // rounds up
```

Comment:
Hi Philippe,
We used ATLAS fully-split xAOD for testing, which is a rather unbalanced TTree, 10K branches, with 8K having baskets smaller than 512 bytes. To achieve good I/O performance ATLAS uses auto-flush 100, resulting in the smallest baskets being ~300-400 bytes, so this change increases the memory by about 8k*150 ~ MB, at the same time it significantly reduces the number of total baskets because it ensures that 100 entries can be stored in a single basket (the old optimization tended to make baskets too small). In ATLAS, where data amount varies event to event, the effect is somewhat less, but in a toy example with fixed sized structures we found a factor of 2 few baskets needed in the new scheme (e.g. 10K entries with auto-flush 100 are store in 100 baskets, not 200).
Thanks, Peter